src/par_iter/range.rs

```diff
@@ -20,44 +19,62 @@ macro_rules! range_impl {
 
         impl ParallelIterator for RangeIter<$t> {
             type Item = $t;
-            type Shared = ();
-            type State = Self;
 
-            fn state(self) -> (Self::Shared, Self::State) {
-                ((), self)
+            fn drive_unindexed<'c, C: UnindexedConsumer<'c, Item=Self::Item>>(self,
+                                                                              consumer: C,
+                                                                              shared: &'c C::Shared)
+                                                                              -> C::Result {
+                bridge(self, consumer, &shared)
             }
         }
 
         unsafe impl BoundedParallelIterator for RangeIter<$t> {
+            fn upper_bound(&mut self) -> usize {
+                ExactParallelIterator::len(self)
+            }
+
+            fn drive<'c, C: Consumer<'c, Item=Self::Item>>(self,
+                                                           consumer: C,
+                                                           shared: &'c C::Shared)
+                                                           -> C::Result {
+                bridge(self, consumer, &shared)
+            }
         }
 
         unsafe impl ExactParallelIterator for RangeIter<$t> {
+            fn len(&mut self) -> usize {
+                self.range.len() as usize
```

Comment:
Yeah, I originally was using `u64` but got sick of casting and decided to go with `usize` for now. Not clear if that is the right call though.